### PR TITLE
refactor shrink_collect

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
ancient append vecs need to be built in 1 shot as opposed to being appended to.

#### Summary of Changes
Refactor `shrink_collect` to allow the collection of unique accounts to occur prior to calling `shrink_collect`

Fixes #

### CODE REVIEW

**Review by apfitzge — DISMISSED** [review]
Only had some stylistic comments that I'm not going to block on - though did have a small comment fix i'd like you to commit!

I think the change looks good, but wondering if we should/need to rethink the design of `shrink_collect`. I understand why we have to pass in unique accounts as a reference, but we have the same code duplicated in several places at this point:
`get_unique_accounts_for_shrink(...) -> shrink_collect(...)`

I can't think of a *simple* way to do this better, so happy to sign off on this.

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -3707,26 +3707,32 @@ impl AccountsDb {
...
+    ) -> GetUniqueAccountsResult<'a> {
+        let (result, storage_read_elapsed) = measure!(self.get_unique_accounts_from_storage(store));
+        stats
+            .storage_read_elapsed
+            .fetch_add(storage_read_elapsed.as_us(), Ordering::Relaxed);
+        result
+    }
+
     /// shared code for shrinking normal slots and combining into ancient append vecs
     /// note 'stored_accounts' is passed by ref so we can return references to data within it, avoiding self-references
```
Comment (apfitzge) [inline runtime/src/accounts_db.rs:3723]:
```suggestion
    /// note 'unique_accounts' is passed by ref so we can return references to data within it, avoiding self-references
```

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -3707,26 +3707,32 @@ impl AccountsDb {
...
 
+    fn get_unique_accounts_from_storage_for_shrink<'a>(
+        &self,
+        store: &'a Arc<AccountStorageEntry>,
+        stats: &ShrinkStats,
+    ) -> GetUniqueAccountsResult<'a> {
+        let (result, storage_read_elapsed) = measure!(self.get_unique_accounts_from_storage(store));
+        stats
+            .storage_read_elapsed
+            .fetch_add(storage_read_elapsed.as_us(), Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3718]: love it. will use it as I go.

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -3839,13 +3845,11 @@ impl AccountsDb {
             // It is 'correct' to ignore calls to shrink when a slot is still in the write cache.
             return;
         }
-        let mut stored_accounts = Vec::default();
+        let unique_accounts =
+            self.get_unique_accounts_from_storage_for_shrink(store, &self.shrink_stats);
         debug!("do_shrink_slot_store: slot: {}", slot);
```
[the remark this answers is not in the document]
Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3850]: > but we have the same code duplicated in several places at this point:

Yes, that is why it used to be shared in `shrink_collect`. But, the algorithm for 1 of the cases, ancient append vecs, will now depart severely from the current code. So, to retain as much shared code as possible, I needed to break out the collection of accounts from the index portions. Thus, this pr.

**Pull request merged**